crates/web/src/lib.rs

```diff
@@ -59,10 +59,216 @@ impl<'a> rose::FuncNode for &'a Func {
+                    rose::Binop::Div => writeln!(&mut s, "x{} / x{}", left.var(), right.var())?,
+                },
+                rose::Expr::Call { func, arg } => {
+                    writeln!(&mut s, "F{}(x{})", func.func(), arg.var())?
+                }
+                rose::Expr::If { cond, then, els } => {
+                    writeln!(&mut s, "if x{} {{", cond.var())?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    let x = def.blocks[then.block()].arg.var();
+                    writeln!(&mut s, "  x{x}: T{}", def.vars[x].ty())?;
+                    print_block(s, def, spaces + 2, *then)?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    writeln!(&mut s, "}} else {{")?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    let y = def.blocks[els.block()].arg.var();
+                    writeln!(&mut s, "  x{y}: T{}", def.vars[y].ty())?;
+                    print_block(s, def, spaces + 2, *els)?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    writeln!(&mut s, "}}")?
+                }
+                rose::Expr::For { index, body } => {
+                    writeln!(
+                        &mut s,
+                        "for x{}: T{} {{",
+                        def.blocks[body.block()].arg.var(),
+                        index.ty()
+                    )?;
+                    print_block(s, def, spaces + 2, *body)?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    writeln!(&mut s, "}}")?
+                }
+                rose::Expr::Accum { var, vector, body } => {
+                    writeln!(&mut s, "accum x{}: T{} {{", var.var(), vector.ty())?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    let x = def.blocks[body.block()].arg.var();
+                    writeln!(&mut s, "  x{x}: T{}", def.vars[x].ty())?;
+                    print_block(s, def, spaces + 2, *body)?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    writeln!(&mut s, "}}")?
+                }
+                rose::Expr::Add { accum, addend } => {
+                    writeln!(&mut s, "x{} += x{}", accum.var(), addend.var())?
+                }
+            }
+        }
+        for _ in 0..spaces {
+            write!(&mut s, " ")?;
+        }
+        writeln!(&mut s, "x{}", def.blocks[b.block()].ret.var())?;
+        Ok(())
+    }
+
+    fn print_elems(s: &mut String, items: &Vec<rose::id::Var>) -> std::fmt::Result {
```

Comment:
In Rust it's usually more idiomatic to take a slice reference instead of a `Vec` reference:

```suggestion
    fn print_elems(s: &mut String, items: &[rose::id::Var]) -> std::fmt::Result {
```

Reply:
Note though that this comment is made obsolete by the below ones about generalizing this function to take an `Iterator`.